File: LeetCode_Solutions/402-remove-k-digits/remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        if(num.length()==k)return "0";
        LinkedList<Character> st=new LinkedList<>();
        StringBuilder sb=new StringBuilder();
        for(int i=0;i<num.length();i++){
            while(k>0&&!st.isEmpty()&&st.getLast()>num.charAt(i)){
                st.pollLast();
                k--;
            }
            st.addLast(num.charAt(i));
        }
        while(k-- > 0){
            st.pollLast();
        }
        for(char c:st){
            if(c=='0'&&sb.length()==0){
                continue;
            }
            sb.append(c);
        }
        return sb.length()==0?"0":sb.toString();
    }
}

File: LeetCode_Solutions/402-remove-k-digits/remove-k-digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        if len(num)==k:
            return '0'
        ans=[]
        st=[]
        for i,digit in enumerate(num):
            while k>0 and st and st[-1]>digit:
                st.pop()
                k-=1
            st.append(digit)
        for i in range(k):
            st.pop()
        for i in st:
            if i=='0' and len(ans)==0:
                continue
            ans.append(i)
        return "".join(ans) if ans else '0'